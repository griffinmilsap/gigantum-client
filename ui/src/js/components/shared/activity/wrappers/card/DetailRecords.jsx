// vendor
import React, { Component } from 'react';
import { QueryRenderer, graphql } from 'react-relay';
import ReactMarkdown from 'react-markdown';
import classNames from 'classnames';
// Components
import CodeBlock from 'Components/labbook/renderers/CodeBlock';
// environment
import environment from 'JS/createRelayEnvironment';
// assets
import './DetailRecords.scss';

const DetailRecordsQuery = graphql`
query DetailRecordsQuery($name: String!, $owner: String!, $keys: [String]){
  labbook(name: $name, owner: $owner){
    id
    description
    detailRecords(keys: $keys){
      id
      action
      key
      data
      type
      show
      importance
      tags
    }
  }
}`;
const DetailRecordsDatasetsQuery = graphql`
query DetailRecordsDatasetsQuery($name: String!, $owner: String!, $keys: [String]){
  dataset(name: $name, owner: $owner){
    id
    description
    detailRecords(keys: $keys){
      id
      action
      key
      data
      type
      show
      importance
      tags
    }
  }
}`;


type Props = {
  keys: Array,
  name: String,
  owner: String,
  sectionType: String,
};

class DetailRecords extends Component<Props> {
  /**
   * Lifecycle methods start
   */
  componentDidMount() {
    const self = this;
    setTimeout(() => {
      self._setLinks();
    }, 100);
    window.addEventListener('resize', this._setLinks.bind(this));
  }

  componentDidUpdate() {
    const self = this;
    setTimeout(() => {
      self._setLinks();
    }, 100);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this._setLinks.bind(this));
  }

  /**
   * Lifecycle methods end
   */
  _setLinks = () => {
    const elements = Array.prototype.slice.call(document.getElementsByClassName('ReactMarkdown'));
    const pElements = Array.prototype.slice.call(document.getElementsByClassName('DetailsRecords__link'));
    const fadeElements = Array.prototype.slice.call(document.getElementsByClassName('DetailsRecords__fadeout'));
    const fadeElementsKeys = Object.keys(fadeElements);
    const pElementKeys = Object.keys(pElements);
    const moreObj = {};

    elements.forEach((elOuter, index) => {
      if (this._checkOverflow(elOuter) === true) {
        moreObj[index] = true;
      }
      if (this._checkOverflow(elOuter.childNodes[elOuter.childNodes.length - 1]) === true) {
        moreObj[index] = true;
      }
    });

    pElementKeys.forEach((key) => {
      if (!moreObj[key]) {
        pElements[key].className = 'DetailsRecords__link hidden';
      } else {
        pElements[key].className = 'DetailsRecords__link';
      }
    });

    fadeElementsKeys.forEach((key) => {
      if (!moreObj[key]) {
        fadeElements[key].className = 'DetailsRecords__fadeout hidden';
      } else {
        fadeElements[key].className = 'DetailsRecords__fadeout';
      }
    });
  }

  /**
    @param {Object} element
    checks if elements scrollheight is greater than it's client height
    @return {boolean} isOverflowing
  */
  _checkOverflow = (element) => {
    if (element) {
      const curOverflow = element.style.overflow;
      if (!curOverflow || (curOverflow === 'visible')) {
        element.style.overflow = 'hidden';
      }

      const isOverflowing = element.clientHeight + 3 < element.scrollHeight;
      return isOverflowing;
    }
  }

  /**
    @param {Array} item
    returns tag to render if item matches a case
    @return {JSX}
  */
  _renderDetail = (item) => {
    switch (item[0]) {
      case 'text/plain':
        return (<div className="ReactMarkdown"><p>{item[1]}</p></div>);
      case 'image/png':
        return (<img alt="detail" src={item[1]} />);
      case 'image/jpg':
        return (<img alt="detail" src={item[1]} />);
      case 'image/jpeg':
        return (<img alt="detail" src={item[1]} />);
      case 'image/bmp':
        return (<img alt="detail" src={item[1]} />);
      case 'image/gif':
        return (<img alt="detail" src={item[1]} />);
      case 'text/markdown':
        return (
          <ReactMarkdown
            renderers={{ code: props => <CodeBlock {...props} /> }}
            className="ReactMarkdown"
            source={item[1]}
          />
        );
      default:
        return (<b>{item[1]}</b>);
    }
  }

  /**
    @param {Object} target
    toggles content in a record when it has more to show
    @return {}
  */
  _moreClicked = (target) => {
    // TODO remove setting classNames in react, using state is preffered
    if (target.className !== 'DetailsRecords__link-clicked') {
      const elements = Array.prototype.slice.call(document.getElementsByClassName('ReactMarkdown'));
      const index = Array.prototype.slice.call(document.getElementsByClassName('DetailsRecords__link')).indexOf(target);
      elements[index].className = 'ReactMarkdown-long';
      target.className = 'DetailsRecords__link-clicked';
      target.textContent = 'Less...';
      target.previousSibling.className = 'hidden';
    } else {
      const elements = Array.prototype.slice.call(document.getElementsByClassName('ReactMarkdown-long'));
      const index = Array.prototype.slice.call(document.getElementsByClassName('DetailsRecords__link-clicked')).indexOf(target);
      elements[index].className = 'ReactMarkdown';
      target.className = 'DetailsRecords__link';
      target.textContent = 'More...';
      target.previousSibling.className = 'DetailsRecords__fadeout';
    }
  }

  render() {
    const {
      keys,
      name,
      owner,
      sectionType,
    } = this.props;
    const variables = {
      keys,
      name,
      owner,
    };

    const query = (sectionType === 'labbook')
      ? DetailRecordsQuery
      : DetailRecordsDatasetsQuery;

    return (
      <QueryRenderer
        environment={environment}
        query={query}
        variables={variables}
        render={(response) => {
          if (response && response.props) {
            return (
              <div className="DetailsRecords">
                <ul className="DetailsRecords__list">
                  { response.props[sectionType].detailRecords.map((detailRecord) => {
                    const isNote = (detailRecord.type === 'NOTE');
                    const liCSS = classNames({
                      'DetailsRecords__item-note': isNote,
                      DetailsRecords__item: !isNote,
                    });
                    const containerCSS = classNames({
                      'DetailsRecords__container note': isNote,
                      DetailsRecords__container: !isNote,
                    });
                    return (
                      <div className={containerCSS} key={detailRecord.id}>
                        {(!isNote)
                          && <div className={`DetailsRecords__action DetailsRecords__action--${detailRecord.action && detailRecord.action.toLowerCase()}`} />
                        }

                        {
                          detailRecord.data
                          && detailRecord.data.map((item, index) => {
                            const type = index === 0
                              ? 'type'
                              : 'data';
                            const key = `${detailRecord.id}_${type}`;
                            return (
                              <li
                                className={liCSS}
                                key={key}
                              >
                                {this._renderDetail(item)}
                                <div className="DetailsRecords hidden" />
                                <p
                                  className="DetailsRecords__link hidden"
                                  onClick={e => this._moreClicked(e.target)}
                                  role="presentation"
                                >
                                  More...
                                </p>
                                {this._setLinks()}
                              </li>
                            );
                          })}
                      </div>
                    );
                  })}
                </ul>
              </div>
            );
          }
          return (
            <div className="DetailsRecords__loader-group">
              <div className="DetailsRecords__loader" />
              <div className="DetailsRecords__loader" />
              <div className="DetailsRecords__loader" />
            </div>
          );
        }}
      />
    );
  }
}

export default DetailRecords;
