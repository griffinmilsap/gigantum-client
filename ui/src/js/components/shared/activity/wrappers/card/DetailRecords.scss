@import '~Styles/_imports.scss';

.DetailsRecords {
  overflow: visible;
}

.LargeIcon {
  background-size: 38px;
}

.SmallIcon {
  background-size: 22px !important;
}

.DetailsRecords__list {
  margin: 0;
  position: relative;
  height: auto;
}

.DetailsRecords__item {
  margin-bottom: 1px;
  padding: 5px 15px 5px 75px;

  img {
    width: -moz-fit-content;
    width: fit-content;

  }

  @include flex(flex-start, column);
}

.DetailsRecords__item-note {
  margin-bottom: 1px;
  border-bottom: 1px solid $lightGrey;
  padding: 0 0 0 15px;

  img {
    width: -moz-fit-content;
    width: fit-content;
  }

  @include flex(flex-start, column);
}

@keyframes fluxuateLoad {
    0% {
      opacity: 0.1;
    }
    50% {
      opacity: 0.3;
    }
    100% {
      opacity: 0.1;
    }
}


.DetailsRecords__loader {
  width: 100%;
}

.DetailsRecords__loader {
  margin: 10px 10px 10px 66px;
  width: calc(100% - 66px);
  height: 2px;
  background-color: $primaryText;
  opacity: 0.1;
  animation: fluxuateLoad 3s infinite;
}

li.DetailsRecords__item *,
li.DetailsRecords__item-note *, {
  white-space: pre;
  max-width: 100%;
  line-height: 24px;
}

li.DetailsRecords__item .ReactMarkdown p,
li.DetailsRecords__item .ReactMarkdown ul,
li.DetailsRecords__item-note .ReactMarkdown p,
li.DetailsRecords__item-note .ReactMarkdown ul, {
  line-height: 24px;
  text-overflow: ellipsis;
  display: -webkit-box;
  white-space: pre-line;
  overflow: hidden;
  word-break: break-word;
  line-clamp: 10;
  -webkit-line-clamp: 10;
  box-orient: vertical;
  -webkit-box-orient: vertical;
}

li.DetailsRecords__item .ReactMarkdown,
li.DetailsRecords__item-note .ReactMarkdown {
  max-height: 400px;
  overflow: hidden;
}

.DetailsRecords__link {
  text-align: right;
  margin-right: 20px;
  color: $primaryText;
}

.DetailsRecords__link:hover {
  color: $secondary;
  cursor: pointer;
}

.DetailsRecords__link-clicked {
  text-align: right;
  margin-right: 20px;
  color: $primaryText;
}

.DetailsRecords__link-clicked:hover {
  color: $secondary;
  cursor: pointer;
}

.DetailsRecords__fadeout {
  position: relative;
  bottom: 2em;
  height: 2em;
  background: -webkit-linear-gradient(
      rgba(255, 255, 255, 0) 0%,
      rgba(255, 255, 255, 1) 100%
  );
}

.DetailsRecords__action{
  position: absolute;
  height: 30px;
  width: 30px;
  top: calc(50% - 15px);
  left: 35px;
  @extend .SmallIcon;
}

.DetailsRecords__action--create{
  background: url('~Images/icons/action-create.svg') no-repeat center center;
}
.DetailsRecords__action--edit{
  background: url('~Images/icons/action-edit.svg') no-repeat center center;
}
.DetailsRecords__action--delete{
  background: url('~Images/icons/action-delete.svg') no-repeat center center;
}
.DetailsRecords__action--execute{
  background: url('~Images/icons/action-run.svg') no-repeat center center;
}
.DetailsRecords__action--noaction{
  background: url('~Images/icons/action-null.svg') no-repeat center center;
}
.DetailsRecords__container {
  position: relative;

  &:before {
    content: "";
    display: block;
    position: absolute;
    z-index: 3;
    border-right: 1px dashed $grey;
    height: calc(100% + 0px);
    width: 1px;
    top: calc(-50%);
    left: 25px;
  }
}

.DetailsRecords__action:before {
  content: "";
  display: block;
  position: absolute;
  z-index: index($elements, card);
  border-bottom: 1px dashed $grey;
  height: 1px;
  width: 14px;
  top: 14px;
  left: -8px
}


.DetailsRecords__container:first-child:before {
  content: "";
  display: block;
  position: absolute;
  z-index: 3;
  border-right: 1px dashed $grey;
  height: calc(50% + 10px);
  width: 1px;
  top: -10px;
  left: 25px;
}

.ActivityDetail__details.note .DetailsRecords__container:first-child:before {
  display: none;
}


div.DetailsRecords__container:last-child .DetailsRecords__item:after {
  content: "";
  display: block;
  position: absolute;
  z-index: 3;
  border-right: 1px dashed $grey;
  height: 100%;
  width: 1px;
  top: -50%;
  left: 25px;
}

div.DetailsRecords__container:first-child .DetailsRecords__item:after {
  content: "";
  display: block;
  position: absolute;
  z-index: 3;
  border-right: 1px dashed $grey;
  height: calc(50% + 10px);
  width: 1px;
  top: -10px;
  left: 25px;
}
