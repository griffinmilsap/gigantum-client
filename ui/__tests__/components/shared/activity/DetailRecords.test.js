// vendor
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import relayTestingUtils from '@gigantum/relay-testing-utils';
// components
import DetailRecords from 'Components/shared/activity/wrappers/card/DetailRecords';
// data
import json from './__relaydata__/DetailRecords.json';


let fixtures = {
  keys: [],
  sectionType: 'labbook',
};

global.data['DetailRecordsQuery'] = json.data

describe('DetailRecords', () => {
  it('Renders a snapshot', () => {
    const wrapper = renderer.create(relayTestingUtils.relayWrap(
       <DetailRecords {...fixtures} />, {}, json.data.labbook)
    );

    const tree = wrapper.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
