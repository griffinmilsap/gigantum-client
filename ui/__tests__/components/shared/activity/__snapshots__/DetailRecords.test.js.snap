// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailRecords Renders a snapshot 1`] = `
<div
  className="DetailsRecords"
>
  <ul
    className="DetailsRecords__list"
  >
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
    <div
      className="DetailsRecords__container"
    >
      <div
        className="DetailsRecords__action DetailsRecords__action--null"
      />
    </div>
  </ul>
</div>
`;
